guides/flow-api/build-a-custom-login-page.mdx

```diff
@@ -0,0 +1,546 @@
+        const actionName = button.dataset.action;
+        await state.actions[actionName].run();
+    });
+  });
+}
+
+function renderError(error) {
+  const app = document.getElementById("app");
+
+  app.innerHTML = `
+    <h1>Error</h1>
+    <p class="error">${error.message}</p>
+    <button id="retry">Try Again</button>`;
+
+  document.getElementById("retry").addEventListener("click", () => {
+    initializeLoginFlow();
+  });
+}
+```
+<Note>
+- The action execution logic is embedded in each state's form or button handlers.
+- The `hanko.onAfterStateChange` event ensures the UI updates after each action, and `hanko.onSessionCreated` redirects
+  to `success.html` on successful login.
+- On `success.html`, the logout button calls `hanko.logout()`, and `hanko.onUserLoggedOut` redirects back to the
+  `index.html`.
+- Validation Errors: Each state renderer checks for `state.actions.<action>.inputs.<input>.error` and displays the
+  message (e.g., invalid passcode).
+- Technical Errors: The error state handler displays `state.error.message`.
+</Note>
+
+## Step 4: Run and Test the Login Page
+
+<Warning>
+Always **use your own email address** in the following steps. The Hanko API sends real emails, so proceed with caution!
+</Warning>
+
+<Steps>
+<Step title="Run the Application">
+- Ensure the Hanko Cloud project is configured as described (App URL: `http://localhost:3000`, no passwords, no 2FA,
+  device trust "Never").
+- Update `API_URL` in `script.js` with your Hanko Cloud project's API endpoint.
+- Rebuild the App:
+  ```bash
+  npm run build
+  ```
+- Run the local server:
+  ```bash
+  npm start
+  ```
+- Open `http://localhost:3000` in your browser.
+</Step>
+
+<Step title="Test the Login Flow">
+- Enter the test user's email.
+- Check your email for the passcode, enter the code, and click submit.
+- If prompted, register a passkey.
+- On success, you'll be redirected to `success.html`.
+- Click Log Out to return to the login page.
+</Step>
+
+<Step title="Troubleshooting">
+- CORS Errors: Ensure the App URL in the Hanko Cloud dashboard matches `http://localhost:3000`.
+- Invalid API URL: Verify the `API_URL` in `script.js`.
+- No Passcode Received: Confirm the test user's email is correct and check your spam folder.
+- Passkey Autofill Not Working: Use a passkey-compatible browser (e.g., Chrome) and ensure the input has
+  `autocomplete="username webauthn"`.
+</Step>
+
+</Steps>
+
+## Conclusion
+
+You've built a working login page that:
+- Uses the FlowAPI and hanko-frontend-sdk to handle email/passcode login.
+- Supports passkey autofill in the `login_init` state.
+- Manages sessions with redirect after login and logout functionality.
+- Handles errors gracefully across all specified states.
+
+<Tip>
+Explore next steps like enhancing the UI with advanced CSS or a framework like Bootstrap, adding support for additional
+states or actions (e.g., password login), or integrating the login page into a larger application.
+</Tip>
+
```

Comment:
SInce this guide focuses on the login flow, I think it would be great to also state that you can also use this same process to create your own registration and profile page.

Also I think it would be great to link to the API reference, so that the user can check there which states are available on each flow.

Reply:
e9751177b96578e5af1b18008e06787149fb1760